### PR TITLE
Issue #3062: Bring back node preview

### PR DESCRIPTION
**Pull request by docwilmot**
Preliminary

### CODE REVIEW

File: core/modules/node/node.pages.inc [hunk 1]
```diff
@@ -458,7 +458,7 @@ function node_form($form, &$form_state, Node $node) {
 function node_form_preview($form, &$form_state) {
   $node = $form_state['node'];
```
Comment (klonos) [inline core/modules/node/node.pages.inc]: Given the line after the next one, does this make any sense?

File: core/modules/node/node.theme.inc [hunk 2]
```diff
@@ -4,6 +4,34 @@
...
+ */
+function theme_node_preview_banner_form($variables) {
+  $form = $variables['form'];
+
+  $output = '<div class="node-node-preview-banner-form">';
+  $output .= '<div class="node-node-preview-banner-form-left">';
+  $output .= backdrop_render($form['preview_node_submit']);
+  $output .= backdrop_render($form['backlink']);
+  $output .= '</div>';
+  $output .= '<div class="node-node-preview-banner-form-right">';
```
Comment (herbdool) [inline core/modules/node/node.theme.inc]: Rename to `node-node-preview-banner-form-submit` and `node-node-preview-banner-form-view-mode`.

**Comment by herbdool** [comment]
I did a code review and looks okay to me. I see that there are 4 tests failing but can't get the zenci details to see which ones.

File: core/modules/node/css/node.preview.css [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  box-sizing: border-box;
+  color: #444b53;
+  background: #fff;
+  -webkit-border-radius: 4px;
+  -moz-border-radius: 4px;
+  border-radius: 4px;
+  will-change: border-color;
+  height: 43px;
+}
+@media (min-width: 34em) {
```
Comment (herbdool) [inline core/modules/node/css/node.preview.css]: These classes in the media query need to be after classes with no media query since the latter will always override the former.

I find it easier to follow when the media queries are divided up and put right after each appropriate class. So you know what's overriding. Though I don't know what's the general practice with Backdrop.

File: core/modules/node/css/node.preview.css [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  div.node-preview-banner-form-view-mode {
+    float: right;
+  }
+  .node-preview-backlink {
+    margin: 10px;
+    display: inline;
+    margin: 10px 0 10px 10px;
+  }
+}
+.page {
```
Comment (herbdool) [inline core/modules/node/css/node.preview.css]: I don't see the equivalent one for `.node-preview-container` height outside the media query.

File: core/modules/node/css/node.preview.css [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+.node-preview-container {
+  font: normal 16px/1.538 "Open Sans", "Lucida Grande", "Lucida Sans Unicode", sans-serif;
+  background-color: #efefef;
+  z-index: 999;
+  width: 100%;
+  padding: 10px;
+  overflow: auto;
+}
+.page.preview-js-processed .node-preview-container {
```
Comment (herbdool) [inline core/modules/node/css/node.preview.css]: This isn't called on the test box. Maybe use `html.js .node-preview-container`?

Comment (docwilmot) [reply core/modules/node/css/node.preview.css]: This works as expected. `.page.preview-js-processed` only exists if JS is on.

File: core/modules/node/css/node.preview.css (hunk 3 above)
Comment (docwilmot) [reply core/modules/node/css/node.preview.css]: I prefer all media queries in one go, as it looks neater to me, and I like to be able to see everything that happens at a particular break point. But if there is consensus, sure I'll change.

File: core/modules/node/css/node.preview.css (hunk 4 above)
Comment (docwilmot) [reply core/modules/node/css/node.preview.css]: Sorry, I dont understand this comment.
Edit: I think youre referring to the lines
```
  .node-preview-container {
    height: 120px;
  }
```
If so, we set a height on fullscreen but on mobile allow the ribbon to use whatever height as the content progressively collapses into stacked DIVs.

File: core/modules/node/node.pages.inc [hunk 6]
```diff
@@ -704,12 +704,7 @@ function node_preview_banner_form_node_submit(array $form, array &$form_state) {
 function node_preview_banner_form_view_mode_submit(array $form, array &$form_state) {
   $node = $form_state['values']['node'];
   $node_tempstore_id = $form_state['values']['node_tempstore_id'];
-  if (isset($node->nid)) {
```
Comment (herbdool) [inline core/modules/node/node.pages.inc]: Just want to confirm if you meant to commit this part. Doesn't seem related to the CSS.

Update: I see now it's a different commit for the test errors.

File: core/modules/node/css/node.preview.css (hunk 5 above)
Comment (herbdool) [reply core/modules/node/css/node.preview.css]: I don't understand. On the text box the ribbon is showing up on the top for me. JS is enabled but the CSS should be putting it on the bottom I assume.

File: core/modules/node/css/node.preview.css (hunk 4 above)
Comment (herbdool) [reply core/modules/node/css/node.preview.css]: I suppose it works that way because the ribbon is on top now?

File: core/modules/node/css/node.preview.css (hunk 5 above)
Comment (docwilmot) [reply core/modules/node/css/node.preview.css]: Browser issue maybe? On Chrome and Edge its on the bottom for me.

prev

Comment (herbdool) [reply core/modules/node/css/node.preview.css]: Wasn't a browser issue. I see that you closed and reopened the test box. I see it on the bottom now.

**Comment by herbdool** [comment]
Some tests still failing? Can't access the details though.

**Comment by docwilmot** [comment]
Yes some test to do with datetime, whichdoesnt happen locally, and shouldnt be related to this PR, since I didnt do anything Date related. I'm assuming a bot-related thing.
If there isnt any other issue feel free to tag complete.

File: core/modules/node/node.module [hunk 7]
```diff
@@ -3696,3 +3731,72 @@ function node_language_delete($language) {
     ->condition('langcode', $language->langcode)
     ->execute();
 }
+
+/**
+ * Get a node from the tempstore.
+ *
+ * @param string $node_tempstore_id
+ *   The machine node_tempstore_id of the node item to load.
+ */
```
Comment (herbdool) [inline core/modules/node/node.module]: Can you add the return to the docblock?

File: core/modules/node/node.module [hunk 8]
```diff
@@ -3696,3 +3731,72 @@ function node_language_delete($language) {
...
+
+  return $caches[$node_tempstore_id];
+}
+
+/**
+ * Store changes to a node or menu item in the temporary store.
+ *
+ * @param Node $item
+ *   The node item to save into tempstore.
+ * @param string $type
```
Comment (herbdool) [inline core/modules/node/node.module]: params in docblock need to match actual params.

File: core/modules/node/node.module [hunk 9]
```diff
@@ -3696,3 +3731,72 @@ function node_language_delete($language) {
...
+    'uid' => $GLOBALS['user']->uid,
+    'updated' => REQUEST_TIME,
+  );
+  // Keep previews for 24 hours.
+  tempstore_set('node_tempstore', $node_tempstore_id, $item, REQUEST_TIME + 86400);
+}
+
+/**
+ * Remove an item from the object cache.
+ */
```
Comment (herbdool) [inline core/modules/node/node.module]: Needs params in docblock

File: core/modules/node/node.module [hunk 10]
```diff
@@ -3696,3 +3731,72 @@ function node_language_delete($language) {
...
+ * @return Node|false
+ *   A fully-populated node entity, or FALSE if the node is not found.
+ */
+function node_tempstore_load($node_tempstore_id) {
+  return node_get_node_tempstore($node_tempstore_id);
+}
+
+/**
+ * Build tempstore ID.
+ */
```
Comment (herbdool) [inline core/modules/node/node.module]: Needs a return in docblock I think.

File: core/modules/node/node.pages.inc [hunk 11]
```diff
@@ -458,6 +513,200 @@ function node_form_submit($form, &$form_state) {
...
+ * Generates a node preview.
+ *
+ * @param $node
+ *   The node to preview.
+ *
+ * @return
+ *   An HTML-formatted string of a node preview.
+ *
+ * @see node_form_build_preview()
+ */
```
Comment (herbdool) [inline core/modules/node/node.pages.inc]: params in docblock should match the function.

File: core/modules/node/node.pages.inc [hunk 12]
```diff
@@ -458,6 +513,200 @@ function node_form_submit($form, &$form_state) {
...
+
+/**
+ * Get the preview form selection box.
+ *
+ * @param $node
+ *   A node entity.
+ *
+ * @return array $form
+ *   The view mode preview selection form.
+ */
```
Comment (herbdool) [inline core/modules/node/node.pages.inc]: params in docblock should match the function.

File: core/modules/node/node.pages.inc [hunk 13]
```diff
@@ -458,6 +513,200 @@ function node_form_submit($form, &$form_state) {
...
+function node_preview_banner_form(array $form, array $form_state, EntityInterface $node, $node_tempstore_id) {
```
Comment (herbdool) [inline core/modules/node/node.pages.inc]: I'm wondering if it needs to be `Node $node` rather than `EntityInterface $node`? I'm not sure.

File: core/modules/node/node.pages.inc [hunk 14]
```diff
@@ -458,6 +513,200 @@ function node_form_submit($form, &$form_state) {
...
+    )
+  );
+  $form['#theme'] = 'node_preview_banner_form';
+
+  return $form;
+}
+
+/**
+ * Submit handler for the node preview view mode selection form.
+ */
```
Comment (herbdool) [inline core/modules/node/node.pages.inc]: Probably needs: `@see node_preview_banner_form()`

File: core/modules/node/node.pages.inc [hunk 15]
```diff
@@ -458,6 +513,200 @@ function node_form_submit($form, &$form_state) {
...
+    }
+  }
+  else {
+    backdrop_goto('node/' . $node->nid);
+  }
+}
+
+/**
+ * Submit handler for the node preview view mode selection form.
+ */
```
Comment (herbdool) [inline core/modules/node/node.pages.inc]: Probably needs: `@see node_preview_banner_form()`?

File: core/modules/node/node.module (hunk 7 above)
Comment (docwilmot) [reply core/modules/node/node.module]: Done

File: core/modules/node/node.module (hunk 8 above)
Comment (docwilmot) [reply core/modules/node/node.module]: Done

File: core/modules/node/node.module (hunk 9 above)
Comment (docwilmot) [reply core/modules/node/node.module]: Done

File: core/modules/node/node.module (hunk 10 above)
Comment (docwilmot) [reply core/modules/node/node.module]: Done

File: core/modules/node/node.pages.inc (hunk 11 above)
Comment (docwilmot) [reply core/modules/node/node.pages.inc]: Done

File: core/modules/node/node.pages.inc (hunk 12 above)
Comment (docwilmot) [reply core/modules/node/node.pages.inc]: Done

File: core/modules/node/node.pages.inc (hunk 13 above)
Comment (docwilmot) [reply core/modules/node/node.pages.inc]: Done

File: core/modules/node/node.pages.inc (hunk 14 above)
Comment (docwilmot) [reply core/modules/node/node.pages.inc]: Done

File: core/modules/node/node.pages.inc (hunk 15 above)
Comment (docwilmot) [reply core/modules/node/node.pages.inc]: Done

File: core/modules/node/node.pages.inc [hunk 16]
```diff
@@ -458,6 +513,208 @@ function node_form_submit($form, &$form_state) {
...
+      'class' => array('node-preview-save-button', 'preview-button'),
+    )
+  );
+  $form['#theme'] = 'node_preview_banner_form';
+
+  return $form;
+}
+
+/**
+ * Submit handler for the node preview view mode selection form.
```
Comment (herbdool) [inline core/modules/node/node.pages.inc]: This should be "Submit handler for the node preview banner form." no?

File: core/modules/node/node.pages.inc [hunk 17]
```diff
@@ -458,6 +513,208 @@ function node_form_submit($form, &$form_state) {
...
+  $build['values']['op'] = t('Save');
+  backdrop_form_submit($type . '_node_form', $build, $node);
+  $errors = form_get_errors();
+  if (!empty($errors)) {
+    foreach ($errors as $field_name => $message) {
+      watchdog('node', '%field: %message', array('%message'=> $message, '%field' => $field_name));
+    }
+  }
+  else {
+    backdrop_goto('node/' . $node->nid);
```
Comment (herbdool) [inline core/modules/node/node.pages.inc]: I haven't tested your function for this, but `node_form_submit()` has more logic about determining if the user has access to view content first before sending them to the node. It seems like someone who doesn't have access to view will see an "access denied" once they save from the preview page.

Comment (docwilmot) [reply core/modules/node/node.pages.inc]: Fixed

File: core/modules/layout/layout.module [hunk 18]
```diff
@@ -1063,17 +1063,36 @@ function layout_get_layout_by_path($path = NULL, $router_item = NULL) {
   $cache = &backdrop_static(__FUNCTION__);
 
   if (empty($cache[$router_item['path']])) {
-    $layouts = layout_load_multiple_by_path($router_item['path'], TRUE);
+    // If the path is a node preview path, check if a custom layout exists for
+    // nodes and use that instead.
+    $data = NULL;
```
Comment (quicksketch) [inline core/modules/layout/layout.module]: I don't like this node-specific logic being embedded directly in layout.module. We might allow it for the time being, but we should put a `@todo` to a current issue for separating this back out. https://github.com/backdrop/backdrop-issues/issues/1528 might be an adequate starting point

File: core/modules/layout/layout.module [hunk 19]
```diff
@@ -1063,17 +1063,36 @@ function layout_get_layout_by_path($path = NULL, $router_item = NULL) {
...
     $selected_layout = NULL;
     foreach ($layouts as $layout) {
       // Contexts must be set before the layout's access may be checked.
       $contexts = $layout->getContexts();
       foreach ($contexts as $context) {
         if (isset($context->position)) {
 
-          // Check for an object loaded by the menu router.
+          // Check for an object loaded by the menu router. Use a preview from
+          // Tempstore if on a node preview page.
```
Comment (quicksketch) [inline core/modules/layout/layout.module]: Not a fan of this sort of cross-module behavior. :cry:

File: core/modules/node/js/node.preview.js [hunk 20]
```diff
@@ -0,0 +1,28 @@
+
+(function ($) {
+
+Backdrop.behaviors.nodePreview = {
+  attach: function (context) {
+    // Insert the preview admin banner into the page before the layout.
+    $('.node-preview-container').each(function() {
+      $(this).insertBefore($('.page').once('preview-js').find('.layout'));
```
Comment (quicksketch) [inline core/modules/node/js/node.preview.js]: Instead of tying to the "layout" class, we could take the second-highest parent and use that. e.g.

`$topParent = $(this).parents(':not(body)').last();`

This is also a slightly unusual use of `.once()`, just to add a class. If we could do the `.once()` on `.node-preview-container`, that would be more inline with most other behavior attachments. e.g.

```
 $('.node-preview-container').once('node-preview').each(function() {
  $topParent = $(this).parents(':not(body)').last();
  $topParent.addClass('preview-js');
  $topParent.on('click.preview', 'a', function (e) {
    ...
  }
});
```

File: core/modules/node/node.module [hunk 21]
```diff
@@ -1826,6 +1843,24 @@ function node_menu() {
...
+ *
+ * @param EntityInterface $node
+ *   The node entity to be previewed.
+ *
+ * @return bool
+ *   TRUE if the user has access to the preview, FALSE otherwise.
+ *
+ * @see node_menu()
+ */
+function node_access_preview(EntityInterface $node) {
```
Comment (quicksketch) [inline core/modules/node/node.module]: As this is made specifically for nodes and not other entities, I'd use `Node` as the typehint.

File: core/modules/node/node.module [hunk 22]
```diff
@@ -3696,3 +3731,81 @@ function node_language_delete($language) {
...
     ->execute();
 }
+
+/**
+ * Get a node from the tempstore.
+ *
+ * @param string $node_tempstore_id
+ *   The tempstore ID of the node item to load.
+ *
+ * @return Node|false
```
Comment (quicksketch) [inline core/modules/node/node.module]: All-caps `FALSE` I think.

File: core/modules/node/node.pages.inc [hunk 23]
```diff
@@ -358,6 +378,16 @@ function node_form($form, &$form_state, Node $node) {
   }
   elseif (isset($_SERVER['HTTP_REFERER'])) {
     $path = $_SERVER['HTTP_REFERER'];
+    // When coming from preview, the $path is set to the preview page.
+    if (strpos($path, 'node/preview')) {
+      if (isset($node->nid)) {
+        $path = 'node/' . $node->nid;
+      }
+      else {
+        // todo store old $_SERVER['HTTP_REFERER'] so we can redirect to that.
```
Comment (quicksketch) [inline core/modules/node/node.pages.inc]: `todo` => `@todo`

File: core/modules/node/node.pages.inc [hunk 24]
```diff
@@ -458,6 +515,209 @@ function node_form_submit($form, &$form_state) {
...
+  $view_mode = empty($_GET['view_mode']) ? 'full' : $_GET['view_mode'];
+
+  // Property so we can manipulate $page in template_preprocess_node.
+  if ($view_mode == 'full' || $view_mode == 'default') {
+    backdrop_set_title($node->title);
+    $node->node_page_title = TRUE;
+  }
+ 
+  $build = array();
+  // @todo all links should be disabled, except for returning to the edit form.
```
Comment (quicksketch) [inline core/modules/node/node.pages.inc]: I think this has been resolved by the current JS.

File: core/modules/node/node.pages.inc [hunk 25]
```diff
@@ -458,6 +515,209 @@ function node_form_submit($form, &$form_state) {
...
+    //'#title_display' => 'invisible',
+    '#options' => $view_mode_options,
+    '#default_value' => $view_mode,
+    '#attributes' => array(
+      'onChange' => 'this.form.submit()',
+      'class' => array('node-preview-display-mode', 'preview-button'),
+    )
+  );
+  // This must be the first submit function so the 'onChange' event in 
+  // calls it before any other submit.$form['view_mode']
```
Comment (quicksketch) [inline core/modules/node/node.pages.inc]: Remove stray `$form['view_mode']`.

File: core/modules/node/node.pages.inc [hunk 26]
```diff
@@ -458,6 +515,209 @@ function node_form_submit($form, &$form_state) {
...
+    '#type' => 'link',
+    '#title' => t('Back to content editing'),
+    '#href' => $url,
+    '#attributes' => array('class' => array('node-preview-backlink')),
+  );
+
+  $form['view_mode'] = array(
+    '#type' => 'select',
+    '#title' => t('Select a display mode'),
+    //'#title_display' => 'invisible',
```
Comment (quicksketch) [inline core/modules/node/node.pages.inc]: Probably remove this if we're not going to use it.

File: core/modules/node/node.pages.inc [hunk 27]
```diff
@@ -458,6 +515,209 @@ function node_form_submit($form, &$form_state) {
...
+
+  $node->changed = REQUEST_TIME;
+ 
+  // Get the view mode to render the preview in.
+  $view_mode = empty($_GET['view_mode']) ? 'full' : $_GET['view_mode'];
+
+  // Property so we can manipulate $page in template_preprocess_node.
+  if ($view_mode == 'full' || $view_mode == 'default') {
+    backdrop_set_title($node->title);
+    $node->node_page_title = TRUE;
```
Comment (quicksketch) [inline core/modules/node/node.pages.inc]: Could we restore `$node->in_preview` to match what D7 did previously? We should also document this property in node.entity.inc.

File: core/modules/node/node.theme.inc [hunk 28]
```diff
@@ -176,7 +204,7 @@ function template_preprocess_node(&$variables) {
   $uri = entity_uri('node', $node);
   $variables['node_url']  = url($uri['path'], $uri['options']);
   $variables['title']     = check_plain($node->title);
-  $variables['page']      = $variables['view_mode'] == 'full' && node_is_page($node);
+  $variables['page']      = ($variables['view_mode'] == 'full' && node_is_page($node)) || (!empty($node->node_page_title));
```
Comment (quicksketch) [inline core/modules/node/node.theme.inc]: Per previous comment, this may be better to use `$node->in_preview` instead of `$node->node_page_title`.

File: core/modules/taxonomy/taxonomy.module [hunk 29]
```diff
@@ -1259,17 +1259,24 @@ function _taxonomy_get_tid_from_term(TaxonomyTerm $term) {
...
+        $label = $tag->label();
+      }
+    }
+    if (!empty($label)) {
+      // Commas and quotes in tag names are special cases, so encode 'em.
+      if (strpos($tag->name, ',') !== FALSE || strpos($tag->name, '"') !== FALSE) {
+        $typed_tags[] = '"' . str_replace('"', '""', $tag->name) . '"';
+      }
+      else {
+        $typed_tags[] = $label;
```
Comment (quicksketch) [inline core/modules/taxonomy/taxonomy.module]: Are these taxonomy changes related to node preview? Seems like this should be a separate cleanup.

File: core/modules/taxonomy/taxonomy.module [hunk 30]
```diff
@@ -1578,7 +1585,14 @@ function taxonomy_term_title(TaxonomyTerm $term) {
...
   foreach ($items as $item) {
-    $tags[$item['tid']] = isset($item['taxonomy_term']) ? $item['taxonomy_term'] : taxonomy_term_load($item['tid']);
+    // $tags[$item['tid']] = isset($item['taxonomy_term']) ? $item['taxonomy_term'] : taxonomy_term_load($item['tid']);
+    $tid = $item['tid'];
+    if ($tid == 'autocreate') {
+      $tags[] = $item;
+    }
+    else {
+      $tags[$tid] = isset($item['taxonomy_term']) ? $item['taxonomy_term'] : taxonomy_term_load($tid);
+    }
```
Comment (quicksketch) [inline core/modules/taxonomy/taxonomy.module]: Same here, seems unrelated to this PR.

**Review by quicksketch — CHANGES_REQUESTED** [review]
Several items of feedback. My only big hangup is the strong cross-connection between Layout and Node that is added here in `layout_get_layout_by_path()`. Let's make sure we outline decoupling these in an issue and include a `@todo`.

There are still a few tests failures that need to be addressed as well.

Let's also add `@since 1.11.0` to all new functions to clarify their creation date.

File: core/modules/layout/layout.module (hunk 18 above)
Comment (docwilmot) [reply core/modules/layout/layout.module]: todo added

File: core/modules/layout/layout.module (hunk 19 above)
Comment (docwilmot) [reply core/modules/layout/layout.module]: agreed. todo added

File: core/modules/node/js/node.preview.js (hunk 20 above)
Comment (docwilmot) [reply core/modules/node/js/node.preview.js]: done

File: core/modules/node/node.module (hunk 21 above)
Comment (docwilmot) [reply core/modules/node/node.module]: done

File: core/modules/node/node.module (hunk 22 above)
Comment (docwilmot) [reply core/modules/node/node.module]: done

File: core/modules/node/node.pages.inc (hunk 23 above)
Comment (docwilmot) [reply core/modules/node/node.pages.inc]: done

File: core/modules/node/node.pages.inc (hunk 27 above)
Comment (docwilmot) [reply core/modules/node/node.pages.inc]: done

File: core/modules/node/node.pages.inc (hunk 24 above)
Comment (docwilmot) [reply core/modules/node/node.pages.inc]: true, removed.

File: core/modules/node/node.pages.inc (hunk 26 above)
Comment (docwilmot) [reply core/modules/node/node.pages.inc]: done

File: core/modules/node/node.pages.inc (hunk 25 above)
Comment (docwilmot) [reply core/modules/node/node.pages.inc]: done

File: core/modules/node/node.theme.inc (hunk 28 above)
Comment (docwilmot) [reply core/modules/node/node.theme.inc]: done

File: core/modules/taxonomy/taxonomy.module (hunk 29 above)
Comment (docwilmot) [reply core/modules/taxonomy/taxonomy.module]: removed.
this was in the original code from Drupal, but was removed before merge. doesnt seem to be missed.

File: core/modules/taxonomy/taxonomy.module (hunk 30 above)
Comment (docwilmot) [reply core/modules/taxonomy/taxonomy.module]: removed.
this was in the original code from Drupal, but was removed before merge. doesnt seem to be missed.

**Review by quicksketch — APPROVED** [review]

**Comment by herbdool** [comment]
@quicksketch looks the PR has thousands of files changed now, including file permissions

**Comment by quicksketch** [comment]
I tried to rebase everything to avoid that massive permissions diff showing up in the PR, but I seem to have broken it. Now I can't reopen the PR or restore the changes. :face_with_head_bandage: I think I'll have to create a new PR.

**Comment by quicksketch** [comment]
Replacement PR (should be identical other than it's rebased and has one last commit to fix the tests) https://github.com/backdrop/backdrop/pull/2290
